### PR TITLE
Fixes Brave ads OnTextLoaded and OnHtmlLoaded are called multiple times for Google Docs, Slides and Sheets

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/13893

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See https://github.com/brave/brave-browser/issues/13893

### CODE REVIEW

File: browser/brave_ads/ads_tab_helper.cc [hunk 1]
```diff
@@ -144,10 +157,6 @@ void AdsTabHelper::DocumentOnLoadCompletedInMainFrame(
     return;
   }
 
-  std::unique_ptr<dom_distiller::SourcePageHandleWebContents> handle =
-      std::make_unique<dom_distiller::SourcePageHandleWebContents>(
-          web_contents(), false);
-
```
Comment (tmancey) [inline browser/brave_ads/ads_tab_helper.cc:150]: not used

File: browser/brave_ads/ads_tab_helper.cc [hunk 2]
```diff
@@ -91,6 +92,12 @@ void AdsTabHelper::OnJavaScriptHtmlResult(base::Value value) {
   std::string html;
   value.GetAsString(&html);
 
+  const uint32_t html_hash = base::FastHash(html);
+  if (html_hash == html_hash_) {
+    return;
+  }
+  html_hash_ = html_hash;
```
Comment (tmancey) [inline browser/brave_ads/ads_tab_helper.cc:99]: compared against a hash to reduce memory usage

File: browser/brave_ads/ads_tab_helper.cc [hunk 3]
```diff
@@ -105,6 +112,12 @@ void AdsTabHelper::OnJavaScriptTextResult(base::Value value) {
   std::string text;
   value.GetAsString(&text);
 
+  const uint32_t text_hash = base::FastHash(text);
+  if (text_hash == text_hash_) {
+    return;
+  }
+  text_hash_ = text_hash;
+
```
Comment (tmancey) [inline browser/brave_ads/ads_tab_helper.cc:120]: compared against a hash to reduce memory usage

File: browser/brave_ads/ads_tab_helper.h [hunk 4]
```diff
@@ -91,6 +91,8 @@ class AdsTabHelper : public content::WebContentsObserver,
   bool is_browser_active_;
   std::vector<GURL> redirect_chain_;
   bool should_process_;
+  uint32_t html_hash_;
```
Comment (aseren) [inline browser/brave_ads/ads_tab_helper.h:94]: Maybe need to set initial values for class members?

**Review by moritzhaller — APPROVED** [review]

File: browser/brave_ads/ads_tab_helper.h (hunk 4 above)
Comment (tmancey) [reply browser/brave_ads/ads_tab_helper.h:94]: ah yes this was a std::string previously so did not need to, I will fix now

**Review by aseren — APPROVED** [review]

**Pull request merged**